File: python/open_guardrail/guards/duplicate_chunk.py
"""Detect duplicate or near-duplicate chunks in RAG-retrieved text."""
from __future__ import annotations

import re
import time
from typing import List, Optional, Tuple

from open_guardrail.core import GuardResult


def _trigram_similarity(a: str, b: str) -> float:
    if a == b:
        return 1.0
    if len(a) < 3 or len(b) < 3:
        return 1.0 if a == b else 0.0
    tri_a = {a[i : i + 3] for i in range(len(a) - 2)}
    tri_b = {b[i : i + 3] for i in range(len(b) - 2)}
    intersection = len(tri_a & tri_b)
    return (2 * intersection) / (len(tri_a) + len(tri_b))


class _DuplicateChunk:
    def __init__(
        self,
        *,
        action: str = "warn",
        similarity_threshold: float = 0.8,
        min_chunk_length: int = 50,
    ) -> None:
        self.name = "duplicate-chunk"
        self.action = action
        self.similarity_threshold = similarity_threshold
        self.min_chunk_length = min_chunk_length

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        chunks = [
            c.strip()
            for c in re.split(r"\n---\n|\n\n|---", text)
            if len(c.strip()) >= self.min_chunk_length
        ]

        max_sim = 0.0
        dup_pair: Optional[Tuple[int, int]] = None

        for i in range(len(chunks)):
            for j in range(i + 1, len(chunks)):
                sim = _trigram_similarity(
                    chunks[i].lower(), chunks[j].lower()
                )
                if sim > max_sim:
                    max_sim = sim
                    dup_pair = (i, j)

        triggered = max_sim >= self.similarity_threshold
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Duplicate chunks (similarity {max_sim:.2f})" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"similarity": max_sim, "chunk_indices": list(dup_pair)} if triggered and dup_pair else None,
        )


def duplicate_chunk(
    *,
    action: str = "warn",
    similarity_threshold: float = 0.8,
    min_chunk_length: int = 50,
) -> _DuplicateChunk:
    return _DuplicateChunk(
        action=action,
        similarity_threshold=similarity_threshold,
        min_chunk_length=min_chunk_length,
    )

File: packages/guards/src/duplicate-chunk.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface DuplicateChunkOptions {
  action: 'block' | 'warn';
  similarityThreshold?: number;
  minChunkLength?: number;
}

function trigramSimilarity(a: string, b: string): number {
  if (a === b) return 1;
  if (a.length < 3 || b.length < 3) return a === b ? 1 : 0;
  const trigramsA = new Set<string>();
  for (let i = 0; i <= a.length - 3; i++) trigramsA.add(a.slice(i, i + 3));
  const trigramsB = new Set<string>();
  for (let i = 0; i <= b.length - 3; i++) trigramsB.add(b.slice(i, i + 3));
  let intersection = 0;
  for (const t of trigramsA) {
    if (trigramsB.has(t)) intersection++;
  }
  return (2 * intersection) / (trigramsA.size + trigramsB.size);
}

export function duplicateChunk(options: DuplicateChunkOptions): Guard {
  const threshold = options.similarityThreshold ?? 0.8;
  const minLen = options.minChunkLength ?? 50;

  return {
    name: 'duplicate-chunk',
    version: '0.1.0',
    description: 'Detects duplicate or near-duplicate chunks in RAG-retrieved text',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const chunks = text
        .split(/\n---\n|\n\n|---/)
        .map((c) => c.trim())
        .filter((c) => c.length >= minLen);

      let maxSim = 0;
      let dupPair: [number, number] | undefined;

      for (let i = 0; i < chunks.length; i++) {
        for (let j = i + 1; j < chunks.length; j++) {
          const sim = trigramSimilarity(
            chunks[i].toLowerCase(),
            chunks[j].toLowerCase(),
          );
          if (sim > maxSim) {
            maxSim = sim;
            dupPair = [i, j];
          }
        }
      }

      const triggered = maxSim >= threshold;

      return {
        guardName: 'duplicate-chunk',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? maxSim : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { similarity: maxSim, chunkIndices: dupPair }
          : undefined,
      };
    },
  };
}
